core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java

```diff
@@ -1094,16 +1039,116 @@ private AnnotatedValueResolver build() {
                 // - QueryParams (or HttpParameters)
                 // - User classes which can be converted by request converter
                 //
-                // So the container type should be 'null'.
-                //
-                types = new SimpleImmutableEntry<>(null, (Class<?>) type);
+                if (unwrappedParameterizedType instanceof Class) {
+                    elementType = (Class<?>) unwrappedParameterizedType;
+                    parameterizedElementType = null;
+                } else if (unwrappedParameterizedType instanceof ParameterizedType) {
+                    parameterizedElementType = (ParameterizedType) unwrappedParameterizedType;
+                    elementType = (Class<?>) parameterizedElementType.getRawType();
+                } else {
+                    throw new IllegalArgumentException("Unsupported parameter type: " +
+                                                       unwrappedParameterizedType.getTypeName());
+                }
             }
 
             return new AnnotatedValueResolver(annotationType, httpElementName, pathVariable, shouldExist,
-                                              shouldWrapValueAsOptional, types.getKey(), types.getValue(),
-                                              defaultValue, description, resolver,
+                                              isOptional, containerType, elementType,
+                                              parameterizedElementType, defaultValue, description, resolver,
                                               beanFactoryId, aggregation);
         }
+
+        @Nullable
+        private Class<?> getContainerType(Type parameterizedType) {
+            final Class<?> rawType = toRawType(parameterizedType);
+            if (pathVariable) {
+                if (Iterable.class.isAssignableFrom(rawType)) {
+                    throw new IllegalArgumentException(
+                            "Container type is not supported for a path variable: " + httpElementName +
+                            " (" + parameterizedType + ')');
+                }
+            }
+
+            if (!supportContainer) {
+                return null;
+            }
+
+            if (rawType == Iterable.class ||
+                rawType == List.class ||
+                rawType == Collection.class) {
+                return ArrayList.class;
+            }
+
+            if (rawType == Set.class) {
```

Comment:
Can we support `Map`, `GenericArrayType` and arrays as we do in the DocService? 🤔 
Supporting `Map` is maybe too much because we have to use varagrs for `expectedParameterizedResultType` though. 🤔

Reply:
Not sure about `Map`, but I think it's a good idea to support arrays. Maybe later in another PR? 😁

Reply:
https://github.com/line/armeria/issues/2790